### ISSUE TITLE
Question about reproducing ImageNet 64x64 + UniPC results in the paper

### ISSUE THREAD

**Issue by bobxu12138** [scxue/DM-NonUniform#3]
Hi, thank you for releasing this code.

I am trying to reproduce the ImageNet 64x64 experiment in your paper. I tried to follow the paper setting as closely as possible.

Right now, my pipeline is:
	•	OpenAI guided-diffusion 64x64 diffusion model + classifier
	•	UniPC sampler
	•	your step_optim.py for optimized timesteps

I can reproduce the uniform-t and uniform-lambda baselines, and the trend looks reasonable. For example, at 5 NFEs, uniform-t is better than uniform-lambda.

But for the optimized schedule, my result is only a little better than uniform-lambda, and still much worse than uniform-t.

I also printed the optimized timesteps and the surrogate objective. The surrogate objective does go down clearly after optimization, but the final timesteps look a bit strange to me, and the final FID is still not as good as expected.

<img width="836" height="108" alt="Image" src="https://github.com/user-attachments/assets/47aa9494-da4f-4fe8-80cb-c382915b0204" />
<img width="848" height="150" alt="Image" src="https://github.com/user-attachments/assets/ac9d2a67-a6ed-4395-96bc-5c922a184fae" />

So I want to ask: is there any trick or important detail for reproducing this ImageNet 64x64 + UniPC experiment?

**Comment by scxue** [comment]
Hi, thanks a lot for trying this out and for the detailed report — this is very helpful.

One thing to note is that the OpenAI guided-diffusion model uses a **cosine noise schedule**. In this parameterization, the log-SNR at the terminal time (i.e., \(T=1\)) is **extremely small**, which can make the optimization over timesteps less well-behaved in practice.

For ImageNet 64×64 with cosine schedules, we recommend **slightly truncating the time horizon**, e.g.:

- use **\(T = 0.9946\)** instead of \(T = 1\)

This avoids the extremely low log-SNR region near \(t=1\), and typically leads to much more stable and effective timestep optimization.

You may also notice that after this change:
- the optimized timesteps become more reasonable,
- and the final FID improves significantly.

Let us know if this resolves the issue or if you still observe discrepancies!

**Comment by bobxu12138** [comment]
Solve it, Thanks a lot !

**Comment by bobxu12138** [comment]
By the way, for the third experiment implementing on EDM model, I follow your suggestion in Issue 2. 
As in NoiseScheduleEDM:
def marginal_log_mean_coeff(self, t):
    return torch.zeros_like(t)

def marginal_alpha(self, t):
    return torch.ones_like(t)

def marginal_std(self, t):
    return t

def marginal_lambda(self, t):
    log_mean_coeff = self.marginal_log_mean_coeff(t)
    log_std = torch.log(self.marginal_std(t))
    return log_mean_coeff - log_std
which seems to indicate alpha(t) = 1,sigma(t) = t

so I modify the method in StepOptim to avoid getting INF values
from
def sigma(self, t):
    return np.sqrt(1 - self.alpha(t) * self.alpha(t))

def lambda_func(self, t):
    return np.log(self.alpha(t) / self.sigma(t))
to
def sigma(self, t):
    t = torch.as_tensor(t, dtype=torch.float64)
    return t.cpu().numpy()

def lambda_func(self, t):
    t = torch.as_tensor(t, dtype=torch.float64)
    t = torch.clamp(t, min=1e-12)
    return (-torch.log(t)).cpu().numpy()

However, after making this change, the final results are still not very good, so I would like to ask whether this modification is correct, and if it is, whether there are any other parts that also need to be changed when adapting StepOptim to the EDM setting. Thank you very much for your help.

**Comment by scxue** [comment]
Have you set T=80 and t=0.002?

**Comment by bobxu12138** [comment]
````text
Yes, I have followed your suggestion in Issue 2 and substituted the corresponding part
```python
class NoiseScheduleEDM:
    def __init__(
            self,
            dtype=torch.float32,
    ):
        """Create a wrapper class for the forward SDE (EDM type).
        """
        self.T = 80
        self.t0 = 2e-3
        self.schedule = 'continuous'
        self.total_N = 1000

    def marginal_log_mean_coeff(self, t):
        """
        Compute log(alpha_t) of a given continuous-time label t in [0, T].
        """
        return torch.zeros_like(t)

    def marginal_alpha(self, t):
        """
        Compute alpha_t of a given continuous-time label t in [0, T].
        """
        return torch.ones_like(t)

    def marginal_std(self, t):
        """
        Compute sigma_t of a given continuous-time label t in [0, T].
        """
        return t

    def marginal_lambda(self, t):
        """
        Compute lambda_t = log(alpha_t) - log(sigma_t) of a given continuous-time label t in [0, T].
        """
        log_mean_coeff = self.marginal_log_mean_coeff(t)
        log_std = torch.log(self.marginal_std(t))
        return log_mean_coeff - log_std

    def inverse_lambda(self, lamb):
        """
        Compute the continuous-time label t in [0, T] of a given half-logSNR lambda_t.
        """
        return torch.exp(-lamb)

    def edm_sigma(self, t):
        return self.marginal_std(t) / self.marginal_alpha(t)

    def edm_inverse_sigma(self, edmsigma):
        sigma = edmsigma
        lambda_t = torch.log(1 / sigma)
        t = self.inverse_lambda(lambda_t)
        return t

class StepOptim(object):
    def __init__(self, ns):
        super().__init__()
        self.ns = ns
        self.T = 80.0  # t_T of diffusion sampling, for VP models, T=1.0; for EDM models, T=80.0

    def alpha(self, t):
        t = torch.as_tensor(t, dtype=torch.float64)
        return self.ns.marginal_alpha(t).numpy()

    def sigma(self, t):
        t = torch.as_tensor(t, dtype=torch.float64)
        return t.cpu().numpy()

    def lambda_func(self, t):
        t = torch.as_tensor(t, dtype=torch.float64)
        t = torch.clamp(t, min=1e-12)
        return (-torch.log(t)).cpu().numpy()

    def H0(self, h):
        return np.exp(h) - 1

    def H1(self, h):
        return np.exp(h) * h - self.H0(h)

    def H2(self, h):
        return np.exp(h) * h * h - 2 * self.H1(h)

    def H3(self, h):
        return np.exp(h) * h * h * h - 3 * self.H2(h)

    def inverse_lambda(self, lamb):
        """
        Compute the continuous-time label t in [0, T] of a given half-logSNR lambda_t.
        """
        return torch.exp(-lamb)

    def edm_sigma(self, t):
        t = torch.as_tensor(t, dtype=torch.float64)
        return self.ns.edm_sigma(t).cpu().numpy()

    def edm_inverse_sigma(self, edm_sigma):
        edm_sigma = torch.as_tensor(edm_sigma, dtype=torch.float64)
        return self.ns.edm_inverse_sigma(edm_sigma).cpu().numpy()

    def sel_lambdas_lof_obj(self, lambda_vec, eps):

        lambda_eps, lambda_T = self.lambda_func(eps).item(), self.lambda_func(self.T).item()
        lambda_vec_ext = np.concatenate((np.array([lambda_T]), lambda_vec, np.array([lambda_eps])))
        N = len(lambda_vec_ext) - 1

        hv = np.zeros(N)
        for i in range(N):
            hv[i] = lambda_vec_ext[i + 1] - lambda_vec_ext[i]
        elv = np.exp(lambda_vec_ext)
        emlv_sq = np.exp(-2 * lambda_vec_ext)
        alpha_vec = 1. / np.sqrt(1 + emlv_sq)
        sigma_vec = 1. / np.sqrt(1 + np.exp(2 * lambda_vec_ext))
        data_err_vec = (sigma_vec ** 1) / alpha_vec
        # for pixel-space diffusion models, we empirically find (sigma_vec**1)/alpha_vec will be better

        truncNum = 3  # For NFEs <= 7, set truncNum = 3 to avoid numerical instability; for NFEs > 7, truncNum = 0
        res = 0.
        c_vec = np.zeros(N)
        for s in range(N):
            if s in [0, N - 1]:
                n, kp = s, 1
                J_n_kp_0 = elv[n + 1] - elv[n]
                res += abs(J_n_kp_0 * data_err_vec[n])
            elif s in [1, N - 2]:
                n, kp = s - 1, 2
                J_n_kp_0 = -elv[n + 1] * self.H1(hv[n + 1]) / hv[n]
                J_n_kp_1 = elv[n + 1] * (self.H1(hv[n + 1]) + hv[n] * self.H0(hv[n + 1])) / hv[n]
                if s >= truncNum:
                    c_vec[n] += data_err_vec[n] * J_n_kp_0
                    c_vec[n + 1] += data_err_vec[n + 1] * J_n_kp_1
                else:
                    res += np.sqrt((data_err_vec[n] * J_n_kp_0) ** 2 + (data_err_vec[n + 1] * J_n_kp_1) ** 2)
            else:
                n, kp = s - 2, 3
                J_n_kp_0 = elv[n + 2] * (self.H2(hv[n + 2]) + hv[n + 1] * self.H1(hv[n + 2])) / (
                            hv[n] * (hv[n] + hv[n + 1]))
                J_n_kp_1 = -elv[n + 2] * (self.H2(hv[n + 2]) + (hv[n] + hv[n + 1]) * self.H1(hv[n + 2])) / (
                            hv[n] * hv[n + 1])
                J_n_kp_2 = elv[n + 2] * (
                            self.H2(hv[n + 2]) + (2 * hv[n + 1] + hv[n]) * self.H1(hv[n + 2]) + hv[n + 1] * (
                                hv[n] + hv[n + 1]) * self.H0(hv[n + 2])) / (hv[n + 1] * (hv[n] + hv[n + 1]))
                if s >= truncNum:
                    c_vec[n] += data_err_vec[n] * J_n_kp_0
                    c_vec[n + 1] += data_err_vec[n + 1] * J_n_kp_1
                    c_vec[n + 2] += data_err_vec[n + 2] * J_n_kp_2
                else:
                    res += np.sqrt((data_err_vec[n] * J_n_kp_0) ** 2 + (data_err_vec[n + 1] * J_n_kp_1) ** 2 + (
                                data_err_vec[n + 2] * J_n_kp_2) ** 2)
        res += sum(abs(c_vec))
        return res

    def get_ts_lambdas(self, N, eps, initType):
        lambda_eps, lambda_T = self.lambda_func(eps).item(), self.lambda_func(self.T).item()

        # constraints
        constr_mat = np.zeros((N, N - 1))
        for i in range(N - 1):
            constr_mat[i][i] = 1.
            constr_mat[i + 1][i] = -1
        lb_vec = np.zeros(N)
        lb_vec[0], lb_vec[-1] = lambda_T, -lambda_eps

        ub_vec = np.zeros(N)
        for i in range(N):
            ub_vec[i] = np.inf
        linear_constraint = LinearConstraint(constr_mat, lb_vec, ub_vec)

        # initial vector
        if initType in ['unif', 'unif_origin']:
            lambda_vec_ext = torch.linspace(lambda_T, lambda_eps, N + 1)
        elif initType in ['unif_t', 'unif_t_origin']:
            t_vec = torch.linspace(self.T, eps, N + 1)
            lambda_vec_ext = self.lambda_func(t_vec)
        elif initType in ['edm', 'edm_origin']:
            rho = 7
            edm_sigma_min, edm_sigma_max = self.edm_sigma(eps).item(), self.edm_sigma(self.T).item()
            edm_sigma_vec = torch.linspace(edm_sigma_max ** (1. / rho), edm_sigma_min ** (1. / rho), N + 1).pow(rho)
            t_vec = self.edm_inverse_sigma(edm_sigma_vec)
            lambda_vec_ext = self.lambda_func(t_vec)
        elif initType in ['quad', 'quad_origin']:
            t_order = 2
            t_vec = torch.linspace(self.T ** (1. / t_order), eps ** (1. / t_order), N + 1).pow(t_order)
            lambda_vec_ext = self.lambda_func(t_vec)
        else:
            print('InitType not found!')
            return

        if initType in ['unif_origin', 'unif_t_origin', 'edm_origin', 'quad_origin']:
            lambda_res = lambda_vec_ext
            t_res = torch.tensor(self.inverse_lambda(lambda_res))
        else:
            lambda_vec_init = np.array(lambda_vec_ext[1:-1])
            res = minimize(self.sel_lambdas_lof_obj, lambda_vec_init, method='trust-constr', args=(eps),
                           constraints=[linear_constraint], options={'verbose': 1})
            lambda_res = torch.tensor(np.concatenate((np.array([lambda_T]), res.x, np.array([lambda_eps]))))
            t_res = torch.tensor(self.inverse_lambda(lambda_res))
        return t_res, lambda_res

**Comment by scxue** [comment]
From what I recall for the EDM experiments:

I used  
`data_err_vec = (sigma_vec ** 1) / alpha_vec`.

For the truncation setting:
- `truncnum = 3` for NFEs = 5, 6  
- `truncnum = 0` for NFEs ≥ 7  

However, I’m not able to verify the exact original configuration at the moment, since it was on a previous company server that I no longer have access to.

Also, do the optimized timesteps look reasonable on your side?

**Comment by bobxu12138** [comment]
The timestep looks reasonable,  but they seem to move the second step much earlier compared with the other baselines
Fid based on 50k images when NFE=5 :
edm, 18.2876
logSNR, 14.4314
optimized,5, 20.0324

<img width="803" height="154" alt="Image" src="https://github.com/user-attachments/assets/eba427e4-34d5-4c84-8463-f8359985f6a0" />

<img width="739" height="113" alt="Image" src="https://github.com/user-attachments/assets/4a453e80-9921-4a4e-8600-2788b164933e" />

<img width="841" height="98" alt="Image" src="https://github.com/user-attachments/assets/3507bd87-fbb5-4add-9ac2-63e89560c602" />

**Comment by scxue** [comment]
It’s possible that the final reported results were not produced with exactly this config. I might have slightly adjusted the `truncnum` setting in the final runs.

Given that the optimized timesteps look not that reasonable and the FID is worse than expected, I would suggest trying change`truncnum`.
